### ISSUE TITLE
Management of issues

### ISSUE THREAD

**Issue by gpetiot** [ocaml-ppx/ocamlformat#318]
What do you think about using the labels [suggested here](https://medium.com/@dave_lunny/sane-github-labels-c5d2e6004b63) ? The combination of status labels, type labels and priority labels looks pretty convenient (and I think status and priority labels are seriously lacking for this project).

**Comment by jberdine** [comment]
It is definitely true that the current labeling situation is wanting, and those look good. (Sometimes I think that priority should be divided into severity and urgency, but that is probably overly complicated.) Perhaps some modification of that statuses would be appropriate. Some issues are freshly created, some are planned to be resolved, some are in progress, some are waiting for some other event, and some are valid but no one can allocate time to them. Highlighting the last category is important, I think, in order to be clear about expectations and to help direct people in the community who want to help to issues that would otherwise hang around.

It might also help to have a status between "up for grabs" and "in progress", called "design" or something, for having some discussion about the plan for implementing a fix for an issue before getting far into code. Sometimes it is most efficient to understand the space for solutions by doing some programming, but other times it can help to have some discussion first.

For types, we would also need one to gather the feedback and suggestions on the style of the output. Some issues are bugs, but many are instead discussion threads to work through trade-offs and other considerations prior to someone submitting a PR implementing their suggested style.

For PRs, a different set of labels makes sense I expect, since we don't have a 1:1 correspondence between PRs and issues, and PRs have a different lifecycle.

Would you like to propose a set of status, type, and priority labels that you think cover what we need with as few / little ambiguity between labels as possible?

**Comment by hhugo** [comment]
should we consider using https://github.com/ocaml-ppx/ocamlformat/projects ?

**Comment by gpetiot** [comment]
I don't know for the project boards, but concerning labels I think the following set makes sense:

**Status (for issues)**
```
To be discussed -> Accepted -> Help wanted -> In progress -> Closed
                 | Rejected                 | On hold
                                            | Blocked
                                            | Completed
```

**Status (for PRs)**
```
To be discussed -> Help wanted -> In progress -> Approved
                                | On hold      | Rejected
                                | Blocked
                                | Completed
                                | Review needed
```

**Type (for both)**
- Bug
- Suggestion
- Maintenance

**Priority (for both)**
- Critical
- High
- Medium
- Low

**Comment by jberdine** [comment]
Let me see if I understand the intended interpretations.

For issues, they are opened and directly go into the "To discuss" state. Then some are "Rejected" because they are by design or unfixable or whatever. The rest are "Accepted" and enter the "Help wanted" state. When someone starts working on an issue, it moves to the "In Progress" state. From there, it might become "Blocked" on something else, or be put "On Hold" if some progress has been made but it is no longer being actively worked on. Eventually, it is "Closed" when a fix is committed. I don't know what "Completed" is for, and it seems that "Accepted" is redundant. Have I understood?

One wrinkle is with style suggestions, where usually the choice between Accepted and Rejected can only really be made through discussion of a PR implementing the suggestion. It might be helpful to make the expected flow for style suggestions explicit. Something like 1. submitter opens issue with description of the suggestion; 2. discussion on what has already been tried, expected difficulties, etc.; 3. submitter submits PR adding option implementing proposed style change, preferably with discussion of testing done (and preferably with commentary / explanation of the improvement on the diff the change has on `make -C test`, or else that work will all be left to the maintainers); 4. iterative discussion refining/polishing the proposal and implementation; 5. merge; 6. optionally discuss changing defaults to enable new style.

For issues, there are times where people working on fixing an issue are waiting for the submitter to answer a question. I think it would help to have an explicit label for this, to make it clear to submitters that progress is blocked on their answer.

For PRs, I don't think I understand the proposed labels as well. Starting out in the "To Discuss" state makes sense (say for a PR containing just a commit adding a test, or maybe even no commits at all if we can do that). I'm not sure about the transition to Help wanted for PRs. I think "In Progress", "On Hold", and "Blocked" make sense too. I guess that "Review Needed" would be added whenever there are commits pushed after the latest review (can this be automated), or are you thinking of something more proactive that submitters would do to ask for a review? I'm not sure of the intention for "Completed", and is "Approved" redundant with the "Approved" response in reviews? And would the "Closed" state of PRs cover the "Rejected" case?

Does it make sense to also have a "Documentation" type? And for issues, some have been essentially "Question"s from users to developers, so it would probably be good to have a type for them too.

I don't have any personal experience with the Projects page, but I have seen some projects where it seemed to be helpful.

**Comment by gpetiot** [comment]
> I don't know what "Completed" is for, and it seems that "Accepted" is redundant. Have I understood?

Initially I thought of Complete as "the contributor has completed his task/implementation and it is up to the issue's submitter to decide whether the problem has been successfully fixed". If so, the issue is closed, otherwise it is back to the previous state.

I agree with your suggestion for 'style suggestion' issues, maybe a label "feedback" could be used for your steps 4 & 6 to discuss the proposed solution.

> For issues, there are times where people working on fixing an issue are waiting for the submitter to answer a question. I think it would help to have an explicit label for this, to make it clear to submitters that progress is blocked on their answer.

I think the label 'Blocked' is appropriate here, as well as tagging the submitter in the discussion.

> I'm not sure about the transition to Help wanted for PRs.

It could be used if the submitter of the PR does not have the time to complete his PR or has given up on his PR, so that anybody can pick it up.

> I guess that "Review Needed" would be added whenever there are commits pushed after the latest review (can this be automated), or are you thinking of something more proactive that submitters would do to ask for a review? I'm not sure of the intention for "Completed", and is "Approved" redundant with the "Approved" response in reviews? And would the "Closed" state of PRs cover the "Rejected" case?

'Review needed' can be used when the submitter has completed its job (yes 'Completed' is redundant) and is waiting for other contibutors (that will eventually merge his PR) to review the code. The PR is 'approved' by the reviewers if it is ready to be merged. Otherwise it will be 'rejected' if the review suggested changes to the code (maybe the name is too harsh), and the state will be 'in progress' again and so on.

> Does it make sense to also have a "Documentation" type? And for issues, some have been essentially "Question"s from users to developers, so it would probably be good to have a type for them too.

Yes I think it makes sense.

**Comment by jberdine** [comment]
> Initially I thought of Complete as "the contributor has completed his 
> task/implementation and it is up to the issue's submitter to decide 
> whether the problem has been successfully fixed". If so, the issue is 
> closed, otherwise it is back to the previous state.

Ah, ok, that makes sense. I wonder though, if it might be more efficient 
to simply optimistically close the issues, which can easily be reopened? 
It is helpful and convenient to track which commits fix which issues 
using the #close N triggers in commit messages, which we would lose with 
a separate Completed state.

> I agree with your suggestion for 'style suggestion' issues, maybe a 
> label "feedback" could be used for your steps 4 & 6 to discuss the 
> proposed solution.

Yes, could be feedback, or maybe "In Progress".

>> For issues, there are times where people working on fixing an issue 
>> are waiting for the submitter to answer a question. I think it would 
>> help to have an explicit label for this, to make it clear to 
>> submitters that progress is blocked on their answer.
>
> I think the label 'Blocked' is appropriate here, as well as tagging 
> the submitter in the discussion.

Ok, it ought to be clear in these cases that it is blocked waiting on 
action from the submitter, so that should work.

>> I'm not sure about the transition to Help wanted for PRs.
>
> It could be used if the submitter of the PR does not have the time to 
> complete his PR or has given up on his PR, so that anybody can pick it 
> up.

Ok, that makes sense. I interpreted the arrows in the diagram as 
indicating that the expected flow as from "To Discuss" to "Help Wanted" 
and from there to e.g. "In Progress". But your explanation makes it seem 
that "Help Wanted" ought to be in the same column as e.g. "Blocked", 
right?

>> I guess that "Review Needed" would be added whenever there are 
>> commits pushed after the latest review (can this be automated), or 
>> are you thinking of something more proactive that submitters would do 
>> to ask for a review? I'm not sure of the intention for "Completed", 
>> and is "Approved" redundant with the "Approved" response in reviews? 
>> And would the "Closed" state of PRs cover the "Rejected" case?
>
> 'Review needed' can be used when the submitter has completed its job 
> (yes 'Completed' is redundant) and is waiting for other contibutors 
> (that will eventually merge his PR) to review the code. The PR is 
> 'approved' by the reviewers if it is ready to be merged. Otherwise it 
> will be 'rejected' if the review suggested changes to the code (maybe 
> the name is too harsh), and the state will be 'in progress' again and 
> so on.

I think we could use the existing "Changes requested" state instead of 
"Rejected"? In general for PRs it would be nice to have as little 
process as possible without things becoming inefficient. There are not 
so many PRs and my impression is that for the most part we keep track of 
their state pretty well. Perhaps we could be better about using "Request 
changes" and "Request review" as often as they make sense, rather than 
leaving things implicit such as what I have found myself doing: "I just 
pushed some commits, and I assume that everyone knows that that means I 
request a review".

**Comment by gpetiot** [comment]
For info, `merlin` uses this set of labels:
- Kind/Bug
- Kind/Docs
- Kind/Feature-request
- Kind/To-discuss
- Status/0-More-info-needed
- Status/0-Triage
- Status/1-Acknowledged
- Status/2-Regression
- Status/3-Fixed-need-test
- Status/4-Fixed
- Status/5-Awaiting-feedback

**Comment by jberdine** [comment]
Those look good, perhaps with `Kind/Style-suggestion` separated out from `Kind/Feature-request`.

We can iterate on this as we gather experience, and the proposals here are certainly better than the current state. So feel free to go ahead and create a new set of labels and start moving the existing issues over to them.

**Comment by jberdine** [comment]
Additionally, could the labelling system and expected issue and pull request lifetimes be documented, as label descriptions and/or something in the wiki?

**Comment by hhugo** [comment]
It would fit well in contributing.md

**Comment by gpetiot** [comment]
Closed with #449
